// @sanity/components/lib/validation/ValidationStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _errorOutlineIcon = _interopRequireDefault(require("part:@sanity/base/error-outline-icon"));

var _checkIcon = _interopRequireDefault(require("part:@sanity/base/check-icon"));

var _reactTippy = require("react-tippy");

var _ValidationList = _interopRequireDefault(require("./ValidationList"));

var _ValidationStatus = _interopRequireDefault(require("./ValidationStatus.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ValidationStatus extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        markers = _this$props.markers,
        _this$props$showSumma = _this$props.showSummary,
        showSummary = _this$props$showSumma === void 0 ? false : _this$props$showSumma,
        _this$props$hideToolt = _this$props.hideTooltip,
        hideTooltip = _this$props$hideToolt === void 0 ? false : _this$props$hideToolt;
    var validation = markers.filter(marker => marker.type === 'validation');
    var errors = validation.filter(marker => marker.level === 'error');
    var warnings = validation.filter(marker => marker.level === 'warning');

    if (errors.length === 0 && warnings.length === 0) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _ValidationStatus.default.ok
      }, /*#__PURE__*/_react.default.createElement(_checkIcon.default, null));
    }

    var errorDef = "error".concat(errors.length === 1 ? '' : 's');
    var warningDef = "warning".concat(warnings.length === 1 ? '' : 's');
    var errorText = errors.length > 0 ? "".concat(errors.length, " validation ").concat(errorDef) : '';
    var warningText = warnings.length > 0 ? "".concat(warnings.length, " ").concat(warningDef) : '';
    var hasBoth = errors.length > 0 && warnings.length > 0;
    var tooltipText = "Found ".concat(errorText, " ").concat(warningText !== '' ? "".concat(hasBoth ? 'and ' : '').concat(warningText) : '');
    var iconClassName = (0, _classnames.default)(_ValidationStatus.default.icon, (errors === null || errors === void 0 ? void 0 : errors.length) < 1 && (warnings === null || warnings === void 0 ? void 0 : warnings.length) > 0 ? _ValidationStatus.default.warning : _ValidationStatus.default.error);

    var TooltipText = () => /*#__PURE__*/_react.default.createElement("div", {
      className: _ValidationStatus.default.tooltip
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: iconClassName
    }, /*#__PURE__*/_react.default.createElement(_errorOutlineIcon.default, null)), /*#__PURE__*/_react.default.createElement("div", {
      className: _ValidationStatus.default.tooltipText
    }, tooltipText));

    return /*#__PURE__*/_react.default.createElement(_reactTippy.Tooltip, {
      tabIndex: 0,
      arrow: true,
      theme: "light",
      className: _ValidationStatus.default.root,
      disabled: hideTooltip,
      html: showSummary ? /*#__PURE__*/_react.default.createElement(TooltipText, null) : /*#__PURE__*/_react.default.createElement(_ValidationList.default, {
        markers: validation,
        kind: "simple"
      })
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _ValidationStatus.default.inner
    }, validation && validation.length > 0 && /*#__PURE__*/_react.default.createElement("div", {
      className: iconClassName
    }, /*#__PURE__*/_react.default.createElement(_errorOutlineIcon.default, null))));
  }

}

exports.default = ValidationStatus;

_defineProperty(ValidationStatus, "defaultProps", {
  markers: []
});
