// @sanity/components/lib/badges/DefaultBadge.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _DefaultBadge = _interopRequireDefault(require("./DefaultBadge.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DefaultBadge extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        color = _this$props.color,
        children = _this$props.children,
        title = _this$props.title;
    return /*#__PURE__*/_react.default.createElement("span", {
      className: _DefaultBadge.default.root,
      "data-color": color,
      title: title
    }, children);
  }

}

exports.default = DefaultBadge;

_defineProperty(DefaultBadge, "propTypes", {
  color: _propTypes.default.oneOf([undefined, 'success', 'warning', 'danger', 'info']),
  children: _propTypes.default.node.isRequired,
  title: _propTypes.default.string
});

_defineProperty(DefaultBadge, "defaultProps", {
  color: undefined,
  title: undefined
});
