// @sanity/default-layout/lib/navbar/toolMenu/ToolSwitcherItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _pluginIcon = _interopRequireDefault(require("part:@sanity/base/plugin-icon"));

var _tooltip = require("../tooltip");

var _ToolSwitcherItem = _interopRequireDefault(require("./ToolSwitcherItem.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TOUCH_DEVICE = ('ontouchstart' in document.documentElement);

function ToolSwitcherItem(props) {
  var selected = props.selected,
      title = props.title,
      icon = props.icon,
      showIcon = props.showIcon,
      showLabel = props.showLabel;
  var Icon = icon || _pluginIcon.default;
  return /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
    className: (0, _classnames.default)(_ToolSwitcherItem.default.root, selected && _ToolSwitcherItem.default.selected),
    content: title,
    disabled: TOUCH_DEVICE || showLabel,
    key: showLabel ? 'tippy' : 'tippy-disabled',
    title: showLabel ? '' : title
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _ToolSwitcherItem.default.inner,
    tabIndex: -1
  }, /*#__PURE__*/_react.default.createElement("div", null, showIcon && /*#__PURE__*/_react.default.createElement("div", {
    className: _ToolSwitcherItem.default.iconContainer
  }, /*#__PURE__*/_react.default.createElement(Icon, null)), showLabel && /*#__PURE__*/_react.default.createElement("div", {
    className: _ToolSwitcherItem.default.toolName
  }, title))));
}

ToolSwitcherItem.defaultProps = {
  selected: false,
  showIcon: true,
  showLabel: true,
  direction: 'horizontal'
};
var _default = ToolSwitcherItem;
exports.default = _default;
