// @sanity/desk-tool/lib/panes/documentPane/editor/HistoryForm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _formBuilder = _interopRequireDefault(require("part:@sanity/form-builder"));

var _HistoryForm = _interopRequireDefault(require("./HistoryForm.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var noop = () => null;

var noopPatchChannel = {
  onPatch: () => noop,
  receivePatches: noop
};

class HistoryForm extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      focusPath: []
    });

    _defineProperty(this, "handleFocus", focusPath => {
      this.setState({
        focusPath
      });
    });
  }

  // eslint-disable-next-line complexity
  render() {
    var _this$props = this.props,
        schema = _this$props.schema,
        schemaType = _this$props.schemaType,
        document = _this$props.document;
    var focusPath = this.state.focusPath;
    return /*#__PURE__*/_react.default.createElement("form", {
      className: _HistoryForm.default.editor
    }, document ? /*#__PURE__*/_react.default.createElement(_formBuilder.default, {
      onBlur: noop,
      onFocus: this.handleFocus,
      focusPath: focusPath,
      readOnly: true,
      schema: schema,
      type: schemaType,
      value: document,
      patchChannel: noopPatchChannel
    }) : /*#__PURE__*/_react.default.createElement("p", null, "There is no data associated with this history event."));
  }

}

exports.default = HistoryForm;

_defineProperty(HistoryForm, "propTypes", {
  schema: _propTypes.default.object.isRequired,
  schemaType: _propTypes.default.shape({
    name: _propTypes.default.string
  }).isRequired,
  document: _propTypes.default.shape({
    _type: _propTypes.default.string
  })
});

_defineProperty(HistoryForm, "defaultProps", {
  document: undefined
});
