// @sanity/components/lib/presence/Avatar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Avatar;

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

var _autoId = require("@reach/auto-id");

var _Avatar = _interopRequireDefault(require("./Avatar.css"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var W = 21;
var H = 21;

function Avatar(_ref) {
  var _ref$borderColor = _ref.borderColor,
      borderColor = _ref$borderColor === void 0 ? 'currentColor' : _ref$borderColor,
      imageUrl = _ref.imageUrl,
      label = _ref.label,
      _ref$isAnimating = _ref.isAnimating,
      isAnimating = _ref$isAnimating === void 0 ? false : _ref$isAnimating,
      children = _ref.children,
      onImageLoadError = _ref.onImageLoadError,
      _ref$position = _ref.position,
      position = _ref$position === void 0 ? 'inside' : _ref$position,
      animateArrowFrom = _ref.animateArrowFrom,
      _ref$status = _ref.status,
      status = _ref$status === void 0 ? 'online' : _ref$status,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'small' : _ref$size,
      tone = _ref.tone;
  var elementId = (0, _autoId.useId)();

  var _useState = (0, _react.useState)(animateArrowFrom || position),
      _useState2 = _slicedToArray(_useState, 2),
      arrowPosition = _useState2[0],
      setArrowPosition = _useState2[1];

  (0, _react.useEffect)(() => {
    var arrowTimer = setTimeout(() => {
      setArrowPosition(position);
    }, 50);
    return () => {
      clearTimeout(arrowTimer);
    };
  }, [position]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.root,
    "data-dock": arrowPosition,
    "data-tone": tone,
    "aria-label": label,
    title: label
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.avatar,
    "data-status": status,
    "data-size": size
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.arrow,
    "data-dock": arrowPosition
  }, /*#__PURE__*/_react.default.createElement("svg", {
    viewBox: "0 0 11 7",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("path", {
    d: "M6.67948 1.50115L11 7L0 7L4.32052 1.50115C4.92109 0.736796 6.07891 0.736795 6.67948 1.50115Z",
    fill: "currentColor"
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.avatarInner
  }, /*#__PURE__*/_react.default.createElement("svg", {
    viewBox: "0 0 ".concat(W, " ").concat(H),
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("ellipse", {
    "data-avatar-border": true,
    cx: W / 2,
    cy: H / 2,
    rx: W / 2 - 1,
    ry: H / 2 - 1,
    transform: "rotate(90 ".concat(W / 2, " ").concat(H / 2, ")"),
    stroke: borderColor,
    className: (0, _classnames.default)(_Avatar.default.border, isAnimating && _Avatar.default.isAnimating)
  }), imageUrl && /*#__PURE__*/_react.default.createElement("ellipse", {
    className: _Avatar.default.backgroundFill,
    "data-avatar-fill": true,
    cx: W / 2,
    cy: H / 2,
    rx: W / 2 - 2,
    ry: H / 2 - 2,
    transform: "rotate(90 ".concat(W / 2, " ").concat(H / 2, ")")
  }), /*#__PURE__*/_react.default.createElement("circle", {
    "data-avatar-image": true,
    cx: W / 2,
    cy: H / 2,
    r: W / 2 - (imageUrl ? 2 : 0),
    fill: imageUrl ? "url(#".concat(elementId, "-image-url)") : 'currentColor'
  }), /*#__PURE__*/_react.default.createElement("defs", null, /*#__PURE__*/_react.default.createElement("pattern", {
    id: "".concat(elementId, "-image-url"),
    patternContentUnits: "objectBoundingBox",
    width: "1",
    height: "1"
  }, imageUrl && /*#__PURE__*/_react.default.createElement("image", {
    href: imageUrl,
    width: "1",
    height: "1",
    onError: () => onImageLoadError(new Error('Image failed to load'))
  })))), children && /*#__PURE__*/_react.default.createElement("div", {
    className: _Avatar.default.avatarInitials
  }, children))));
}
