// @sanity/components/lib/presence/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Avatar", {
  enumerable: true,
  get: function get() {
    return _Avatar.default;
  }
});
Object.defineProperty(exports, "AvatarProvider", {
  enumerable: true,
  get: function get() {
    return _UserAvatar.default;
  }
});
Object.defineProperty(exports, "PopoverList", {
  enumerable: true,
  get: function get() {
    return _PopoverList.default;
  }
});
Object.defineProperty(exports, "StackCounter", {
  enumerable: true,
  get: function get() {
    return _StackCounter.default;
  }
});
Object.defineProperty(exports, "GlobalPresence", {
  enumerable: true,
  get: function get() {
    return _GlobalPresence.GlobalPresence;
  }
});
Object.defineProperty(exports, "PresenceOverlay", {
  enumerable: true,
  get: function get() {
    return _PresenceOverlay.PresenceOverlay;
  }
});
Object.defineProperty(exports, "FieldPresence", {
  enumerable: true,
  get: function get() {
    return _FieldPresence.FieldPresence;
  }
});
Object.defineProperty(exports, "PresenceRegion", {
  enumerable: true,
  get: function get() {
    return _PresenceOverlayRegion.PresenceRegion;
  }
});
Object.defineProperty(exports, "Context", {
  enumerable: true,
  get: function get() {
    return _context.Context;
  }
});
Object.defineProperty(exports, "PresenceScope", {
  enumerable: true,
  get: function get() {
    return _PresenceScope.PresenceScope;
  }
});

var _Avatar = _interopRequireDefault(require("./Avatar"));

var _UserAvatar = _interopRequireDefault(require("./UserAvatar"));

var _PopoverList = _interopRequireDefault(require("./PopoverList"));

var _StackCounter = _interopRequireDefault(require("./StackCounter"));

var _GlobalPresence = require("./GlobalPresence");

var _PresenceOverlay = require("./overlay/PresenceOverlay");

var _FieldPresence = require("./FieldPresence");

var _PresenceOverlayRegion = require("./overlay/PresenceOverlayRegion");

var _context = require("./context");

var _PresenceScope = require("./PresenceScope");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
