// @sanity/components/lib/snackbar/SnackbarItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _circleCheckIcon = _interopRequireDefault(require("part:@sanity/base/circle-check-icon"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _errorIcon = _interopRequireDefault(require("part:@sanity/base/error-icon"));

var _infoIcon = _interopRequireDefault(require("part:@sanity/base/info-icon"));

var _SnackbarItem = _interopRequireDefault(require("./styles/SnackbarItem.css"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class SnackbarItem extends _react.default.Component {
  constructor(props, context) {
    super(props, context);

    _defineProperty(this, "DEFAULT_ICONS", {
      info: /*#__PURE__*/_react.default.createElement(_infoIcon.default, null),
      success: /*#__PURE__*/_react.default.createElement(_circleCheckIcon.default, null),
      warning: /*#__PURE__*/_react.default.createElement(_warningIcon.default, null),
      error: /*#__PURE__*/_react.default.createElement(_errorIcon.default, null)
    });

    _defineProperty(this, "snackIcon", () => {
      var _this$props = this.props,
          icon = _this$props.icon,
          kind = _this$props.kind;
      if (typeof icon === 'boolean' && icon) return this.DEFAULT_ICONS[kind];
      if (typeof icon === 'object' || typeof icon === 'string') return icon;
      return undefined;
    });

    _defineProperty(this, "handleAutoDismissSnack", () => {
      var _this$props2 = this.props,
          autoDismissTimeout = _this$props2.autoDismissTimeout,
          isPersisted = _this$props2.isPersisted,
          id = _this$props2.id,
          onDismiss = _this$props2.onDismiss,
          onClose = _this$props2.onClose;

      if (!isPersisted) {
        this._dismissTimer = setTimeout(() => {
          if (onClose) onClose();
          onDismiss(id);
        }, autoDismissTimeout);
      }
    });

    _defineProperty(this, "handleMouseOver", () => {
      this.cancelAutoDismissSnack();
    });

    _defineProperty(this, "handleMouseLeave", () => {
      var isPersisted = this.props.isPersisted;

      if (!isPersisted) {
        this.handleAutoDismissSnack();
      }
    });

    _defineProperty(this, "handleFocus", () => {
      this.cancelAutoDismissSnack();
    });

    _defineProperty(this, "handleBlur", () => {
      var isPersisted = this.props.isPersisted;

      if (!isPersisted) {
        this.handleAutoDismissSnack();
      }
    });

    _defineProperty(this, "handleAction", () => {
      var _this$props3 = this.props,
          action = _this$props3.action,
          id = _this$props3.id,
          onDismiss = _this$props3.onDismiss;
      if (action && action.callback) action.callback();
      return onDismiss(id);
    });

    _defineProperty(this, "handleClose", () => {
      var _this$props4 = this.props,
          id = _this$props4.id,
          onClose = _this$props4.onClose,
          onDismiss = _this$props4.onDismiss;
      if (onClose) onClose();
      return onDismiss(id);
    });

    _defineProperty(this, "cancelAutoDismissSnack", () => {
      clearTimeout(this._dismissTimer);
    });

    this.state = {
      isEntering: true
    };
    this._snackRef = /*#__PURE__*/_react.default.createRef();
  }

  componentDidMount() {
    var _this$props5 = this.props,
        onSetHeight = _this$props5.onSetHeight,
        id = _this$props5.id,
        isPersisted = _this$props5.isPersisted,
        setAutoFocus = _this$props5.setAutoFocus;

    if (setAutoFocus) {
      this._snackRef.current.focus();

      this.cancelAutoDismissSnack();
    }

    var height = this._snackRef.current && this._snackRef.current.clientHeight;
    onSetHeight(id, height);
    if (isPersisted) this.cancelAutoDismissSnack();else this.handleAutoDismissSnack();
    this._enterTimer = setTimeout(() => {
      this.setState({
        isEntering: false
      });
    }, 100);
  }

  componentWillUnmount() {
    clearTimeout(this._dismissTimer);
    clearTimeout(this._enterTimer);
  }

  render() {
    var _this$props6 = this.props,
        action = _this$props6.action,
        children = _this$props6.children,
        icon = _this$props6.icon,
        id = _this$props6.id,
        isCloseable = _this$props6.isCloseable,
        isOpen = _this$props6.isOpen,
        kind = _this$props6.kind,
        title = _this$props6.title,
        subtitle = _this$props6.subtitle,
        offset = _this$props6.offset;
    var rootStyles = this.state.isEntering ? "".concat(_SnackbarItem.default.root) : "".concat(_SnackbarItem.default.root, " ").concat(isOpen ? _SnackbarItem.default.showSnack : _SnackbarItem.default.dismissSnack);
    var transition = "all 200ms ease-in-out";

    var role = () => {
      if (kind === 'success') return 'status';
      if (kind === 'info') return 'log';
      return 'alert';
    };

    return /*#__PURE__*/_react.default.createElement("div", {
      "aria-label": kind,
      "aria-describedby": "snackbarTitle-".concat(kind, "-").concat(id),
      role: role(),
      ref: this._snackRef,
      tabIndex: "-1",
      className: rootStyles,
      style: {
        bottom: offset,
        transition: transition
      },
      onMouseOver: () => this.handleMouseOver(),
      onMouseLeave: () => this.handleMouseLeave(),
      onFocus: () => this.handleFocus(),
      onBlur: () => this.handleBlur(),
      onKeyDown: e => e.key === 'escape' && this.handleAction(),
      "data-kind": kind
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _SnackbarItem.default.inner
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _SnackbarItem.default.buttonsWrapper
    }, action && /*#__PURE__*/_react.default.createElement("div", {
      className: _SnackbarItem.default.actionButtonContainer
    }, /*#__PURE__*/_react.default.createElement(_default.default // eslint-disable-next-line react/jsx-handler-names
    , {
      onClick: () => this.handleAction(),
      bleed: true,
      color: "white",
      kind: "simple",
      padding: "none"
    }, action.title)), isCloseable && /*#__PURE__*/_react.default.createElement("div", {
      className: _SnackbarItem.default.closeButtonContainer
    }, /*#__PURE__*/_react.default.createElement(_default.default, {
      "aria-label": "Close",
      onClick: this.handleClose,
      bleed: true,
      color: "white",
      icon: _closeIcon.default,
      kind: "simple",
      padding: "none"
    }))), icon && /*#__PURE__*/_react.default.createElement("div", {
      role: "img",
      "aria-hidden": true,
      className: _SnackbarItem.default.icon
    }, this.snackIcon()), /*#__PURE__*/_react.default.createElement("div", {
      className: _SnackbarItem.default.content
    }, /*#__PURE__*/_react.default.createElement("div", {
      id: "snackbarTitle-".concat(kind, "-").concat(id),
      className: _SnackbarItem.default.title
    }, title), subtitle && /*#__PURE__*/_react.default.createElement("div", {
      className: _SnackbarItem.default.subtitle
    }, subtitle), children && /*#__PURE__*/_react.default.createElement("div", {
      className: _SnackbarItem.default.children
    }, children))));
  }

}

exports.default = SnackbarItem;

_defineProperty(SnackbarItem, "propTypes", {
  action: _propTypes.default.shape({
    title: _propTypes.default.string,
    icon: _propTypes.default.node,
    callback: _propTypes.default.func
  }),
  autoDismissTimeout: _propTypes.default.number,
  children: _propTypes.default.node,
  icon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node, _propTypes.default.bool]),
  isCloseable: _propTypes.default.bool,
  isOpen: _propTypes.default.bool.isRequired,
  isPersisted: _propTypes.default.bool,
  id: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]).isRequired,
  kind: _propTypes.default.oneOf(['info', 'warning', 'error', 'success']),
  title: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),
  subtitle: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),
  onDismiss: _propTypes.default.func,
  offset: _propTypes.default.number,
  onClose: _propTypes.default.func,
  onSetHeight: _propTypes.default.func,
  setAutoFocus: _propTypes.default.bool
});

_defineProperty(SnackbarItem, "defaultProps", {
  action: undefined,
  autoDismissTimeout: 5000,
  children: null,
  icon: null,
  isCloseable: true,
  isPersisted: false,
  kind: 'info',
  offset: null,
  onDismiss: () => {},
  onSetHeight: () => {},
  setAutoFocus: false
});
