// @sanity/components/lib/presence/stories/circle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CircleStory = CircleStory;

var _react = _interopRequireDefault(require("react"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _PresenceCircle = _interopRequireDefault(require("../PresenceCircle"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function CircleStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      fontSize: '2em'
    }
  }, /*#__PURE__*/_react.default.createElement(_PresenceCircle.default, {
    text: (0, _knobs.text)('text', 'KG', 'props'),
    title: (0, _knobs.text)('title', undefined, 'props'),
    animateOnHover: (0, _knobs.boolean)('animateOnHover', false, 'props'),
    interactive: (0, _knobs.boolean)('interactive', false, 'props'),
    imageUrl: (0, _knobs.text)('imageUrl', 'https://placeimg.com/64/64', 'props'),
    color: (0, _knobs.color)('color', '#D0021B', 'props')
  }));
}
