// @sanity/components/lib/dialogs/stories/dialogContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogContentStory = DialogContentStory;

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _default = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _content = _interopRequireDefault(require("part:@sanity/components/dialogs/content"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DialogContentStory() {
  var actions = [{
    index: '1',
    title: 'Finish',
    color: 'primary',
    autoFocus: true
  }, {
    index: '2',
    title: 'Cancel'
  }, {
    index: '3',
    title: 'Secondary',
    color: 'danger',
    secondary: true
  }];
  var dialogActions = (0, _knobs.boolean)('Show actions', false, 'test') ? actions : [];
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    title: (0, _knobs.text)('title', undefined, 'dialog props'),
    color: (0, _knobs.select)('color', ['default', 'danger', 'success', 'info', 'warning'], undefined, 'dialog props'),
    onAction: (0, _actions.action)('onAction'),
    actions: dialogActions
  }, /*#__PURE__*/_react.default.createElement(_content.default, {
    size: (0, _knobs.select)('size', ['default', 'small', 'medium', 'large', 'auto'], 'default', 'dialogcontent props'),
    padding: (0, _knobs.select)('padding', ['none', 'small', 'medium', 'large'], 'medium', 'dialogcontent props')
  }, (0, _knobs.text)('content', 'This is the raw content. use DialogContent to size it', 'props'))));
}
