// @sanity/default-layout/lib/navbar/actionButton/ActionButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ActionModal = _interopRequireDefault(require("./ActionModal"));

var _fab = _interopRequireDefault(require("part:@sanity/components/buttons/fab"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ActionButton extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    this.handleToggleModal = this.handleToggleModal.bind(this);
    this.state = {
      modalOpen: false
    };
  }

  handleToggleModal() {
    this.setState(state => ({
      modalOpen: !state.modalOpen
    }));
  }

  render() {
    if (this.state.modalOpen) {
      return /*#__PURE__*/_react.default.createElement(_ActionModal.default, {
        actions: this.props.actions,
        onClose: this.handleToggleModal
      });
    }

    return /*#__PURE__*/_react.default.createElement(_fab.default, {
      onClick: this.handleToggleModal,
      colored: true
    });
  }

}

var _default = ActionButton;
exports.default = _default;
