// @sanity/form-builder/lib/inputs/PortableText/Objects/BlockObject.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockObject = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _portableTextEditor = require("@sanity/portable-text-editor");

var _paths = require("@sanity/util/paths");

var _BlockObjectPreview = require("./BlockObjectPreview");

var _BlockObject = _interopRequireDefault(require("./BlockObject.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/prop-types */
var BlockObject = (_ref) => {
  var _ref$attributes = _ref.attributes,
      focused = _ref$attributes.focused,
      selected = _ref$attributes.selected,
      path = _ref$attributes.path,
      editor = _ref.editor,
      markers = _ref.markers,
      onFocus = _ref.onFocus,
      readOnly = _ref.readOnly,
      type = _ref.type,
      value = _ref.value;
  var validation = markers.filter(marker => marker.type === 'validation');
  var errors = validation.filter(marker => marker.level === 'error');
  var classnames = (0, _classnames.default)([_BlockObject.default.root, focused && _BlockObject.default.focused, selected && _BlockObject.default.selected, errors.length > 0 && _BlockObject.default.hasErrors]);

  var handleClickToOpen = event => {
    if (focused) {
      event.preventDefault();
      event.stopPropagation();
      onFocus(path.concat(_paths.FOCUS_TERMINATOR));
    } else {
      onFocus(path);
    }
  };

  var handleEdit = () => {
    onFocus(path.concat(_paths.FOCUS_TERMINATOR));
  };

  var handleDelete = () => {
    _portableTextEditor.PortableTextEditor.delete(editor, {
      focus: {
        path,
        offset: 0
      },
      anchor: {
        path,
        offset: 0
      }
    }, {
      mode: 'block'
    });

    _portableTextEditor.PortableTextEditor.focus(editor);
  };

  return /*#__PURE__*/_react.default.createElement("div", {
    className: classnames,
    onDoubleClick: handleClickToOpen
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _BlockObject.default.previewContainer,
    style: readOnly ? {
      cursor: 'default'
    } : {}
  }, /*#__PURE__*/_react.default.createElement(_BlockObjectPreview.BlockObjectPreview, {
    type: type,
    value: value,
    path: path,
    readOnly: readOnly,
    onFocus: onFocus,
    onClickingDelete: handleDelete,
    onClickingEdit: handleEdit
  })));
};

exports.BlockObject = BlockObject;
