// @sanity/default-layout/lib/navbar/Navbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _configSanity = _interopRequireDefault(require("config:sanity"));

var _composeIcon = _interopRequireDefault(require("part:@sanity/base/compose-icon"));

var _hamburgerIcon = _interopRequireDefault(require("part:@sanity/base/hamburger-icon"));

var _toolSwitcher = _interopRequireDefault(require("part:@sanity/default-layout/tool-switcher"));

var _searchIcon = _interopRequireDefault(require("part:@sanity/base/search-icon"));

var _router = require("part:@sanity/base/router");

var sidecar = _interopRequireWildcard(require("part:@sanity/default-layout/sidecar?"));

var _spaces = require("../util/spaces");

var _Branding = _interopRequireDefault(require("./branding/Branding"));

var _LoginStatus = _interopRequireDefault(require("./loginStatus/LoginStatus"));

var _SanityStatusContainer = _interopRequireDefault(require("./studioStatus/SanityStatusContainer"));

var _GlobalPresenceStatus = require("./studioStatus/GlobalPresenceStatus");

var _SearchContainer = _interopRequireDefault(require("./search/SearchContainer"));

var _DatasetSelect = _interopRequireDefault(require("./datasetSelect/DatasetSelect"));

var _tooltip = require("./tooltip");

var _Navbar = _interopRequireDefault(require("./Navbar.css"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TOUCH_DEVICE = ('ontouchstart' in document.documentElement);
var isSidecarEnabled;
var SidecarToggleButton;

if (sidecar) {
  isSidecarEnabled = sidecar.isSidecarEnabled;
  SidecarToggleButton = sidecar.SidecarToggleButton;
}

function Navbar(props) {
  var searchIsOpen = props.searchIsOpen,
      onCreateButtonClick = props.onCreateButtonClick,
      onToggleMenu = props.onToggleMenu,
      onSwitchTool = props.onSwitchTool,
      onUserLogout = props.onUserLogout,
      onSearchOpen = props.onSearchOpen,
      onSearchClose = props.onSearchClose,
      onSetLoginStatusElement = props.onSetLoginStatusElement,
      onSetSearchElement = props.onSetSearchElement,
      router = props.router,
      tools = props.tools,
      user = props.user,
      showLabel = props.showLabel,
      showToolSwitcher = props.showToolSwitcher;
  var searchClassName = _Navbar.default.search;
  if (searchIsOpen) searchClassName += " ".concat(_Navbar.default.searchIsOpen);
  var className = _Navbar.default.root;
  if (showToolSwitcher) className += " ".concat(_Navbar.default.withToolSwitcher);
  var rootState = _spaces.HAS_SPACES && router.state.space ? {
    space: router.state.space
  } : {};
  return /*#__PURE__*/_react.default.createElement("div", {
    className: className,
    "data-search-open": searchIsOpen
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.hamburger
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: _Navbar.default.hamburgerButton,
    type: "button",
    title: "Open menu",
    onClick: onToggleMenu
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.hamburgerButtonInner,
    tabIndex: -1
  }, /*#__PURE__*/_react.default.createElement(_hamburgerIcon.default, null)))), /*#__PURE__*/_react.default.createElement(_router.StateLink, {
    state: rootState,
    className: _Navbar.default.branding
  }, /*#__PURE__*/_react.default.createElement(_Branding.default, {
    projectName: _configSanity.default && _configSanity.default.project.name
  })), _spaces.HAS_SPACES && /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.datasetSelect
  }, /*#__PURE__*/_react.default.createElement(_DatasetSelect.default, {
    isVisible: showToolSwitcher,
    tone: "navbar"
  })), /*#__PURE__*/_react.default.createElement("button", {
    className: _Navbar.default.createButton,
    onClick: onCreateButtonClick,
    type: "button"
  }, /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
    disabled: TOUCH_DEVICE,
    content: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Create new document")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.createButtonInner,
    tabIndex: -1
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.createButtonIcon
  }, /*#__PURE__*/_react.default.createElement(_composeIcon.default, null)), /*#__PURE__*/_react.default.createElement("span", {
    className: _Navbar.default.createButtonText
  }, "New")))), /*#__PURE__*/_react.default.createElement("div", {
    className: searchClassName,
    ref: onSetSearchElement
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_SearchContainer.default, {
    shouldBeFocused: searchIsOpen,
    onOpen: onSearchOpen,
    onClose: onSearchClose
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.toolSwitcher
  }, /*#__PURE__*/_react.default.createElement(_toolSwitcher.default, {
    direction: "horizontal",
    isVisible: showToolSwitcher,
    tools: tools,
    activeToolName: router.state.tool,
    onSwitchTool: onSwitchTool,
    router: router,
    showLabel: showLabel,
    tone: "navbar"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.extras
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.sanityStatus
  }, /*#__PURE__*/_react.default.createElement(_SanityStatusContainer.default, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.presenceStatus
  }, /*#__PURE__*/_react.default.createElement(_GlobalPresenceStatus.GlobalPresenceStatus, null)), isSidecarEnabled && isSidecarEnabled() && /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.sidecarStatus
  }, /*#__PURE__*/_react.default.createElement(SidecarToggleButton, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.loginStatus,
    ref: onSetLoginStatusElement
  }, /*#__PURE__*/_react.default.createElement(_LoginStatus.default, {
    onLogout: onUserLogout,
    user: user
  })), /*#__PURE__*/_react.default.createElement("button", {
    className: _Navbar.default.searchButton,
    onClick: onSearchOpen,
    type: "button"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.searchButtonInner,
    tabIndex: -1
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _Navbar.default.searchButtonIcon
  }, /*#__PURE__*/_react.default.createElement(_searchIcon.default, null)), /*#__PURE__*/_react.default.createElement("span", {
    className: _Navbar.default.searchButtonText
  }, "Search"))));
}

Navbar.defaultProps = {
  showLabel: true,
  showToolSwitcher: true,
  onSetLoginStatusElement: undefined,
  onSetSearchElement: undefined
};
var _default = Navbar;
exports.default = _default;
