// @sanity/form-builder/lib/inputs/PortableText/BlockExtras.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _blockMarkers = _interopRequireDefault(require("part:@sanity/form-builder/input/block-editor/block-markers"));

var _BlockExtras = _interopRequireDefault(require("./BlockExtras.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class BlockExtras extends _react.default.PureComponent {
  getValidationMarkers() {
    var markers = this.props.markers;
    var validation = markers.filter(mrkr => mrkr.type === 'validation');
    return validation.map(mrkr => {
      if (mrkr.path.length <= 1) {
        return mrkr;
      }

      var level = mrkr.level === 'error' ? 'errors' : 'warnings';
      return _objectSpread(_objectSpread({}, mrkr), {}, {
        item: mrkr.item.cloneWithMessage("Contains ".concat(level))
      });
    });
  }

  render() {
    var _this$props = this.props,
        blockActions = _this$props.blockActions,
        markers = _this$props.markers,
        onFocus = _this$props.onFocus,
        renderCustomMarkers = _this$props.renderCustomMarkers,
        isFullscreen = _this$props.isFullscreen;
    var scopedValidation = this.getValidationMarkers();
    var errors = scopedValidation.filter(mrkr => mrkr.level === 'error');
    var warnings = scopedValidation.filter(mrkr => mrkr.level === 'warning');
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)([_BlockExtras.default.root, isFullscreen && _BlockExtras.default.hasFullScreen, errors.length > 0 && _BlockExtras.default.withError, warnings.length > 0 && !errors.length && _BlockExtras.default.withWarning])
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _BlockExtras.default.content
    }, markers.length > 0 && /*#__PURE__*/_react.default.createElement("div", {
      className: _BlockExtras.default.markers
    }, /*#__PURE__*/_react.default.createElement(_blockMarkers.default, {
      className: _BlockExtras.default.markers,
      markers: markers,
      scopedValidation: scopedValidation,
      onFocus: onFocus,
      renderCustomMarkers: renderCustomMarkers
    })), blockActions && /*#__PURE__*/_react.default.createElement("div", {
      className: _BlockExtras.default.blockActions
    }, blockActions)));
  }

}

exports.default = BlockExtras;

_defineProperty(BlockExtras, "defaultProps", {
  blockActions: null,
  renderCustomMarkers: null
});
