// @sanity/form-builder/lib/inputs/PortableText/InvalidValue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _Warning = _interopRequireDefault(require("../Warning"));

var _UnknownFields = _interopRequireDefault(require("../ObjectInput/styles/UnknownFields.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class InvalidValue extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleAction", () => {
      var resolution = this.props.resolution;

      if (resolution) {
        var patches = resolution.patches;
        this.props.onChange({
          type: 'mutation',
          patches
        });
      }
    });

    _defineProperty(this, "handleIgnore", () => {
      this.props.onIgnore();
    });
  }

  render() {
    var resolution = this.props.resolution;

    var message = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", null, resolution.description), /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("pre", {
      className: _UnknownFields.default.inspectValue
    }, JSON.stringify(resolution.item, null, 2))), resolution.action && /*#__PURE__*/_react.default.createElement("div", {
      className: _UnknownFields.default.buttonWrapper
    }, /*#__PURE__*/_react.default.createElement(_default.default, {
      color: "danger",
      onClick: this.handleAction
    }, resolution.action), /*#__PURE__*/_react.default.createElement(_default.default, {
      color: "secondary",
      onClick: this.handleIgnore
    }, "Ignore")));

    return /*#__PURE__*/_react.default.createElement(_Warning.default, {
      heading: "Invalid portable text value",
      message: message
    });
  }

}

exports.default = InvalidValue;
