// @sanity/components/lib/lists/CreateDocument.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _CreateDocument = _interopRequireDefault(require("./styles/CreateDocument.css"));

var _createDocument = _interopRequireDefault(require("part:@sanity/components/previews/create-document"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function CreateDocumentList(props) {
  var _props$items = props.items,
      items = _props$items === void 0 ? [] : _props$items;
  return /*#__PURE__*/_react.default.createElement("ul", {
    className: _CreateDocument.default.root
  }, items.map(choice => /*#__PURE__*/_react.default.createElement("li", {
    key: choice.key,
    className: _CreateDocument.default.item
  }, /*#__PURE__*/_react.default.createElement(_createDocument.default, _extends({}, choice, {
    // eslint-disable-next-line react/jsx-handler-names
    onClick: choice.onClick
  })))));
}

CreateDocumentList.propTypes = {
  items: _propTypes.default.arrayOf(_propTypes.default.shape({
    key: _propTypes.default.string.isRequired,
    title: _propTypes.default.string,
    subtitle: _propTypes.default.string,
    icon: _propTypes.default.func,
    onClick: _propTypes.default.func
  }))
};
var _default = CreateDocumentList;
exports.default = _default;
