// @sanity/components/lib/lists/sortable-grid/SortableGridList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SortableGridList;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _GridList = _interopRequireDefault(require("../grid/GridList"));

var _sortableFactories = require("../sortable-factories");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var Sortable = (0, _sortableFactories.createSortableList)(_GridList.default);

function SortableGridList(props) {
  return /*#__PURE__*/_react.default.createElement(Sortable, _extends({}, props, {
    axis: "xy",
    distance: 1,
    lockToContainerEdges: true,
    transitionDuration: 600
  }));
}

SortableGridList.proTypes = {
  className: _propTypes.default.string,
  movingItemClass: _propTypes.default.string,
  onSortStart: _propTypes.default.func,
  // () => void,
  onSortEnd: _propTypes.default.func // ({oldIndex: number, newIndex: number}) => void

};
