// @sanity/default-layout/lib/navbar/drawer/SideMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _signOutIcon = _interopRequireDefault(require("part:@sanity/base/sign-out-icon"));

var _toolSwitcher = _interopRequireDefault(require("part:@sanity/default-layout/tool-switcher"));

var _DatasetSelect = _interopRequireDefault(require("../datasetSelect/DatasetSelect"));

var _spaces = require("../../util/spaces");

var _SideMenu = _interopRequireDefault(require("./SideMenu.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function SideMenu(props) {
  var activeToolName = props.activeToolName,
      isOpen = props.isOpen,
      onClose = props.onClose,
      onSignOut = props.onSignOut,
      onSwitchTool = props.onSwitchTool,
      router = props.router,
      tools = props.tools,
      user = props.user;
  var className = _SideMenu.default.root;
  if (isOpen) className += " ".concat(_SideMenu.default.isOpen);
  var tabIndex = isOpen ? 0 : -1;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: className
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("button", {
    className: _SideMenu.default.closeButton,
    type: "button",
    onClick: onClose,
    title: "Close menu",
    tabIndex: tabIndex
  }, /*#__PURE__*/_react.default.createElement(_closeIcon.default, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.header
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.userProfile
  }, user.profileImage ? /*#__PURE__*/_react.default.createElement("img", {
    className: _SideMenu.default.userProfileImage,
    src: user.profileImage,
    alt: "".concat(user.name, "'s profile image"),
    "data-initials": user.name ? user.name.charAt(0) : user.email.charAt(0)
  }) : /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.userProfileImageMissing
  }, user.name ? user.name.charAt(0) : user.email.charAt(0)), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.userProfileText
  }, user.name || user.email)), _spaces.HAS_SPACES && /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.datasetSelectContainer
  }, /*#__PURE__*/_react.default.createElement(_DatasetSelect.default, {
    isVisible: isOpen
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.toolSwitcher
  }, /*#__PURE__*/_react.default.createElement(_toolSwitcher.default, {
    activeToolName: activeToolName,
    direction: "vertical",
    isVisible: isOpen,
    onSwitchTool: onSwitchTool,
    router: router,
    tools: tools
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.menuBottom
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SideMenu.default.signOutButton
  }, /*#__PURE__*/_react.default.createElement("a", {
    onClick: onSignOut,
    tabIndex: tabIndex
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _SideMenu.default.signOutButtonIconContainer
  }, /*#__PURE__*/_react.default.createElement(_signOutIcon.default, null)), /*#__PURE__*/_react.default.createElement("span", {
    className: _SideMenu.default.signOutButtonText
  }, "Sign out"))))));
}

SideMenu.defaultProps = {
  activeToolName: null
};
var _default = SideMenu;
exports.default = _default;
