// @sanity/block-tools/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.randomKey = exports.normalizeBlock = exports.getBlockContentFeatures = exports.blocksToEditorValue = exports.editorValueToBlocks = exports.htmlToBlocks = exports.EDITOR_DEFAULT_BLOCK_TYPE = exports.default = void 0;

var _blockContentTypeFeatures = _interopRequireDefault(require("./util/blockContentTypeFeatures"));

var _blocksToEditorValue2 = _interopRequireDefault(require("./converters/blocksToEditorValue"));

var _editorValueToBlocks2 = _interopRequireDefault(require("./converters/editorValueToBlocks"));

var _HtmlDeserializer = _interopRequireDefault(require("./HtmlDeserializer"));

var _constants = require("./constants");

var _normalizeBlock2 = _interopRequireDefault(require("./util/normalizeBlock"));

var _randomKey2 = _interopRequireDefault(require("./util/randomKey"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * BlockTools - various tools for Sanity block content
 *
 * @param {Object} blockContentType
 *    The compiled schema for the block content type to work with
 *
 */
var blockContentFunctions = {
  /**
   * Convert HTML to blocks respecting the block content type's schema
   *
   * @param {String} html
   *
   * @param {Object} options
   *   @property {Object} blockContentType
   *      A compiled version of the schema type for the block content
   *   @property {Array} rules
   *      Optional rules working on the HTML (will be ruled first)
   *   @property {Function} parseHtml
   *      API compatible model as returned from DOMParser for using server side.
   * @returns {Array} Blocks
   */
  htmlToBlocks(html, blockContentType) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    var deserializer = new _HtmlDeserializer.default(blockContentType, options);
    return deserializer.deserialize(html).map(_normalizeBlock2.default);
  },

  /**
   * Convert a serialized editor value to blocks
   *
   * @param {Object} An object representing the structure of the editor value.
   * @param {Object} blockContentType
   * @returns {Array} Blocks
   */
  editorValueToBlocks(value, blockContentType) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    return (0, _editorValueToBlocks2.default)(value, blockContentType, options);
  },

  /**
   * Convert blocks to a serialized editor value
   *
   * @param {Array} blocks
   * @param {Object} blockContentType
   * @returns {Object} An object representing the serialized editor value.
   */
  blocksToEditorValue(blocks, blockContentType) {
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    return (0, _blocksToEditorValue2.default)(blocks, blockContentType, options);
  },

  /**
   * Returns the feature-set of a compiled block content type.
   *
   * @param {Object} blockContentType
   * @returns {Object} The feature-set
   */
  getBlockContentFeatures(blockContentType) {
    return (0, _blockContentTypeFeatures.default)(blockContentType);
  },

  randomKey(length) {
    return (0, _randomKey2.default)(length);
  }

};
var _default = blockContentFunctions;
exports.default = _default;
var EDITOR_DEFAULT_BLOCK_TYPE = _constants.SLATE_DEFAULT_BLOCK;
exports.EDITOR_DEFAULT_BLOCK_TYPE = EDITOR_DEFAULT_BLOCK_TYPE;
var htmlToBlocks = blockContentFunctions.htmlToBlocks;
exports.htmlToBlocks = htmlToBlocks;
var editorValueToBlocks = blockContentFunctions.editorValueToBlocks;
exports.editorValueToBlocks = editorValueToBlocks;
var blocksToEditorValue = blockContentFunctions.blocksToEditorValue;
exports.blocksToEditorValue = blocksToEditorValue;
var getBlockContentFeatures = blockContentFunctions.getBlockContentFeatures;
exports.getBlockContentFeatures = getBlockContentFeatures;
var normalizeBlock = _normalizeBlock2.default;
exports.normalizeBlock = normalizeBlock;
var randomKey = _randomKey2.default;
exports.randomKey = randomKey;
