// @sanity/components/lib/previews/CreateDocumentPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _CreateDocumentPreview = _interopRequireDefault(require("./styles/CreateDocumentPreview.css"));

var _router = require("part:@sanity/base/router");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var fieldProp = _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node, _propTypes.default.func]);

class CreateDocumentPreview extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        title = _this$props.title,
        subtitle = _this$props.subtitle,
        _this$props$media = _this$props.media,
        media = _this$props$media === void 0 ? this.props.icon : _this$props$media,
        isPlaceholder = _this$props.isPlaceholder,
        mediaDimensions = _this$props.mediaDimensions,
        description = _this$props.description,
        params = _this$props.params,
        templateParams = _this$props.templateParams;

    if (isPlaceholder || !params) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _CreateDocumentPreview.default.placeholder
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: _CreateDocumentPreview.default.heading
      }, /*#__PURE__*/_react.default.createElement("h2", {
        className: _CreateDocumentPreview.default.title
      }, "Loading\u2026"), /*#__PURE__*/_react.default.createElement("h3", {
        className: _CreateDocumentPreview.default.subtitle
      }, "Loading\u2026")), media !== false && /*#__PURE__*/_react.default.createElement("div", {
        className: _CreateDocumentPreview.default.media
      }));
    }

    return /*#__PURE__*/_react.default.createElement(_router.IntentLink, {
      intent: "create",
      params: [params, templateParams],
      className: _CreateDocumentPreview.default.root,
      title: subtitle ? "Create new ".concat(title, " (").concat(subtitle, ")") : "Create new ".concat(title),
      onClick: this.props.onClick
    }, media !== false && /*#__PURE__*/_react.default.createElement("div", {
      className: _CreateDocumentPreview.default.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", {
      className: _CreateDocumentPreview.default.mediaString
    }, media), /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement("div", {
      className: _CreateDocumentPreview.default.heading
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _CreateDocumentPreview.default.title
    }, typeof title !== 'function' && title, typeof title === 'function' && title({
      layout: 'default'
    })), subtitle && /*#__PURE__*/_react.default.createElement("h3", {
      className: _CreateDocumentPreview.default.subtitle
    }, typeof subtitle === 'function' && subtitle({
      layout: 'default'
    }) || subtitle)), description && /*#__PURE__*/_react.default.createElement("p", {
      className: _CreateDocumentPreview.default.description
    }, description));
  }

}

_defineProperty(CreateDocumentPreview, "propTypes", {
  title: fieldProp,
  subtitle: fieldProp,
  description: fieldProp,
  media: fieldProp,
  icon: _propTypes.default.func,
  isPlaceholder: _propTypes.default.bool,
  params: _propTypes.default.shape({
    template: _propTypes.default.string
  }),
  templateParams: _propTypes.default.object,
  onClick: _propTypes.default.func,
  mediaDimensions: _propTypes.default.shape({
    width: _propTypes.default.number,
    height: _propTypes.default.number,
    fit: _propTypes.default.oneOf(['clip', 'crop', 'fill', 'fillmax', 'max', 'scale', 'min']),
    aspect: _propTypes.default.number
  })
});

_defineProperty(CreateDocumentPreview, "defaultProps", {
  title: 'Untitled',
  subtitle: undefined,
  params: undefined,
  icon: undefined,
  media: undefined,
  mediaDimensions: {
    width: 80,
    height: 80,
    aspect: 1,
    fit: 'crop'
  }
});

var _default = CreateDocumentPreview;
exports.default = _default;
