// @sanity/components/lib/previews/InlinePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _inlineStyle = _interopRequireDefault(require("part:@sanity/components/previews/inline-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var fieldProp = _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node, _propTypes.default.func]);

class InlinePreview extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        title = _this$props.title,
        media = _this$props.media,
        mediaDimensions = _this$props.mediaDimensions,
        children = _this$props.children;

    if (!title && !media) {
      return /*#__PURE__*/_react.default.createElement("span", null);
    }

    return /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.root
    }, media && /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media !== 'function' && media, /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.title
    }, typeof title === 'function' && title({
      layout: 'inline'
    }) || title), children && /*#__PURE__*/_react.default.createElement("span", {
      className: _inlineStyle.default.children
    }, children));
  }

}

exports.default = InlinePreview;

_defineProperty(InlinePreview, "propTypes", {
  title: fieldProp,
  media: fieldProp,
  children: _propTypes.default.node,
  mediaDimensions: _propTypes.default.shape({
    width: _propTypes.default.number,
    height: _propTypes.default.number,
    fit: _propTypes.default.oneOf(['clip', 'crop', 'fill', 'fillmax', 'max', 'scale', 'min']),
    aspect: _propTypes.default.number
  })
});

_defineProperty(InlinePreview, "defaultProps", {
  title: undefined,
  children: undefined,
  media: undefined,
  mediaDimensions: {
    width: 32,
    height: 32,
    fit: 'crop',
    aspect: 1
  }
});
