// @sanity/default-layout/lib/navbar/studioStatus/SanityStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SanityStatus;

var _react = _interopRequireDefault(require("react"));

var _packageIcon = _interopRequireDefault(require("part:@sanity/base/package-icon"));

var _autoId = require("@reach/auto-id");

var _CurrentVersionsDialog = _interopRequireDefault(require("./CurrentVersionsDialog"));

var _UpdateNotifierDialog = _interopRequireDefault(require("./UpdateNotifierDialog"));

var _SanityStatus = _interopRequireDefault(require("./SanityStatus.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function formatUpdateLabel(len) {
  if (len === 1) return ' 1 update';
  return "".concat(len, " updates");
}

function SanityStatus(props) {
  var isSupported = props.isSupported,
      isUpToDate = props.isUpToDate,
      level = props.level,
      onHideDialog = props.onHideDialog,
      onShowDialog = props.onShowDialog,
      outdated = props.outdated,
      showDialog = props.showDialog,
      versions = props.versions;
  var elementId = (0, _autoId.useId)();
  var currentLevel = outdated.length ? level : 'notice';
  var severity = isSupported ? currentLevel : 'high';
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _SanityStatus.default.root
  }, showDialog && /*#__PURE__*/_react.default.createElement("div", {
    role: "dialog",
    "aria-modal": "true",
    "aria-labelledby": elementId
  }, isUpToDate ? /*#__PURE__*/_react.default.createElement(_CurrentVersionsDialog.default, {
    onClose: onHideDialog,
    versions: versions
  }) : /*#__PURE__*/_react.default.createElement(_UpdateNotifierDialog.default, {
    onClose: onHideDialog,
    outdated: outdated,
    severity: severity
  })), !isUpToDate && /*#__PURE__*/_react.default.createElement("button", {
    className: _SanityStatus.default.button,
    onClick: onShowDialog,
    type: "button",
    "aria-label": "".concat(formatUpdateLabel(outdated.length), ", ").concat(severity, " severity level."),
    id: elementId
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SanityStatus.default.buttonInner,
    tabIndex: -1
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SanityStatus.default.hasUpdates
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _SanityStatus.default.updateIcon,
    role: "image"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _SanityStatus.default.updateIndicator,
    "data-severity": severity,
    "aria-label": "".concat(formatUpdateLabel(outdated.length), ", ").concat(severity, " severity level.")
  }), /*#__PURE__*/_react.default.createElement(_packageIcon.default, null))))));
}

SanityStatus.defaultProps = {
  outdated: []
};
