// @sanity/default-layout/lib/navbar/loginStatus/LoginStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactClickOutside = _interopRequireDefault(require("react-click-outside"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/menus/default"));

var _signOutIcon = _interopRequireDefault(require("part:@sanity/base/sign-out-icon"));

var _LoginStatus = _interopRequireDefault(require("./LoginStatus.css"));

var _reactTippy = require("react-tippy");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class LoginStatus extends _react.default.PureComponent {
  constructor(props) {
    super(props);

    _defineProperty(this, "handleUserMenuItemClick", item => {
      var onLogout = this.props.onLogout;

      if (item.action === 'signOut') {
        onLogout();
      }
    });
  }

  render() {
    var user = this.props.user;

    if (!user) {
      return null;
    }

    var className = _LoginStatus.default.root;
    if (this.props.className) className += this.props.className;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: className
    }, /*#__PURE__*/_react.default.createElement(_reactTippy.Tooltip, {
      trigger: "click",
      interactive: true,
      arrow: true,
      distance: 1,
      theme: "light",
      html: /*#__PURE__*/_react.default.createElement("div", {
        className: _LoginStatus.default.menuWrapper
      }, /*#__PURE__*/_react.default.createElement(_default2.default, {
        onAction: this.handleUserMenuItemClick,
        items: [{
          title: "Sign out",
          icon: _signOutIcon.default,
          action: 'signOut'
        }],
        origin: "top-right",
        onClickOutside: () => {}
      }))
    }, /*#__PURE__*/_react.default.createElement("button", {
      className: _LoginStatus.default.button,
      title: "Show user menu",
      type: "button"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _LoginStatus.default.inner,
      tabIndex: -1
    }, user.profileImage ? /*#__PURE__*/_react.default.createElement("img", {
      src: user.profileImage,
      className: _LoginStatus.default.userImage,
      alt: "".concat(user.name, "'s profile image"),
      "data-initials": (user.name || user.email || '?').charAt(0)
    }) : /*#__PURE__*/_react.default.createElement("div", {
      className: _LoginStatus.default.userImageMissing
    }, user.name ? user.name.charAt(0) : user.email.charAt(0))))));
  }

}

_defineProperty(LoginStatus, "defaultProps", {
  className: undefined,
  onLogout: undefined,
  user: undefined
});

var _default = (0, _reactClickOutside.default)(LoginStatus);

exports.default = _default;
