// @sanity/default-layout/lib/stories/searchFieldMobile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchFieldMobileStory = SearchFieldMobileStory;

var _react = _interopRequireDefault(require("react"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _SearchField = _interopRequireDefault(require("../navbar/search/SearchField"));

var _SearchResults = _interopRequireDefault(require("../navbar/search/SearchResults"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function SearchFieldMobileStory() {
  var hasResults = (0, _knobs.boolean)('hasResults', false, 'props');
  var items = hasResults ? ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'] : [];
  var query = (0, _knobs.text)('query', '', 'props');
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: '#444',
      height: '100vh'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: '#fff',
      margin: '0 auto',
      position: 'relative'
    }
  }, /*#__PURE__*/_react.default.createElement(_SearchField.default, {
    isBleeding: true,
    isFocused: (0, _knobs.boolean)('isFocused', false, 'props'),
    isOpen: (0, _knobs.boolean)('isOpen', false, 'props'),
    results: /*#__PURE__*/_react.default.createElement(_SearchResults.default, {
      activeIndex: (0, _knobs.number)('activeIndex', -1, 'props'),
      isLoading: (0, _knobs.boolean)('isLoading', false, 'props'),
      items: items,
      query: query,
      renderItem: key => /*#__PURE__*/_react.default.createElement("div", {
        key: key,
        style: {
          padding: '0.75em 1em'
        }
      }, key)
    }),
    value: query,
    onChange: () => (0, _actions.action)('onChange')
  })));
}
