// @sanity/default-layout/lib/stories/toolSwitcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToolSwitcherStory = ToolSwitcherStory;

var _react = _interopRequireDefault(require("react"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _pluginIcon = _interopRequireDefault(require("part:@sanity/base/plugin-icon"));

var _viewColumnIcon = _interopRequireDefault(require("part:@sanity/base/view-column-icon"));

var _ToolSwitcherWidget = _interopRequireDefault(require("../navbar/toolMenu/ToolSwitcherWidget"));

var _ToolSwitcherItem = _interopRequireDefault(require("../navbar/toolMenu/ToolSwitcherItem"));

var _Navbar = _interopRequireDefault(require("../navbar/Navbar.css"));

var _DefaultLayout = _interopRequireDefault(require("../DefaultLayout.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ToolSwitcherStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _DefaultLayout.default.navBar
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.toolSwitcher
  }, /*#__PURE__*/_react.default.createElement(_ToolSwitcherWidget.default, {
    onSwitchTool: event => (0, _actions.action)('onSwitchTool'),
    tools: [{
      name: 'Desk tool',
      icon: _viewColumnIcon.default
    }, {
      name: 'Plugin 1',
      icon: _pluginIcon.default
    }, {
      name: 'Plugin 2',
      icon: _pluginIcon.default
    }],
    renderItem: tool => /*#__PURE__*/_react.default.createElement(_ToolSwitcherItem.default, {
      title: tool.name,
      icon: tool.icon
    })
  }))));
}
