// @sanity/components/lib/utilities/Escapable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function createListener(eventName) {
  var listeners = [];
  return listen;

  function notify(event) {
    listeners.forEach(l => l(event));
  }

  function unlisten(listener) {
    listeners = listeners.filter(l => l !== listener);

    if (listeners.length === 0) {
      document.removeEventListener(eventName, notify);
    }
  }

  function listen(listener) {
    if (listeners.length === 0) {
      document.addEventListener(eventName, notify);
    }

    listeners.push(listener);
    return () => unlisten(listener);
  }
}

var onKeypress = createListener('keydown');

class Escapable extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleKeyPress", event => {
      if (this.props.onEscape && event.key === 'Escape') {
        this.props.onEscape(event);
      }
    });
  }

  UNSAFE_componentWillMount() {
    this.removeListener = onKeypress(this.handleKeyPress);
  }

  componentWillUnmount() {
    this.removeListener();
  }

  render() {
    return this.props.children || null;
  }

}

exports.default = Escapable;

_defineProperty(Escapable, "propTypes", {
  onEscape: _propTypes.default.func,
  children: _propTypes.default.node
});

_defineProperty(Escapable, "defaultProps", {
  onEscape: () => {},
  children: undefined
});
