// @sanity/components/lib/utilities/ScrollContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ScrollContainer = _interopRequireDefault(require("./styles/ScrollContainer.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ScrollContainer extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleScroll", event => {
      this.props.onScroll(event);
    });

    _defineProperty(this, "setScrollContainerElement", element => {
      this._scrollContainerElement = element;
    });
  }

  getChildContext() {
    return {
      getScrollContainer: () => this._scrollContainerElement
    };
  }

  componentDidMount() {
    if (this.props.onScroll) {
      this._scrollContainerElement.addEventListener('scroll', this.handleScroll, {
        passive: true
      });
    }
  }

  componentWillUnmount() {
    if (this.props.onScroll) {
      this._scrollContainerElement.removeEventListener('scroll', this.handleScroll, {
        passive: true
      });
    }
  }

  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      ref: this.setScrollContainerElement,
      className: "".concat(_ScrollContainer.default.scrollContainer, " ").concat(this.props.className),
      tabIndex: this.props.tabIndex
    }, this.props.children);
  }

}

exports.default = ScrollContainer;

_defineProperty(ScrollContainer, "propTypes", {
  className: _propTypes.default.string,
  children: _propTypes.default.node,
  onScroll: _propTypes.default.func,
  tabIndex: _propTypes.default.number
});

_defineProperty(ScrollContainer, "defaultProps", {
  children: undefined,
  className: '',
  onScroll: () => null,
  tabIndex: undefined
});

_defineProperty(ScrollContainer, "defaultProps", {
  className: ''
});

_defineProperty(ScrollContainer, "childContextTypes", {
  getScrollContainer: _propTypes.default.func
});
