// @sanity/components/lib/utilities/Stacked.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createStack = createStack;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _nanoPubsub = _interopRequireDefault(require("nano-pubsub"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function createStack() {
  var stack = [];
  var changes = (0, _nanoPubsub.default)();

  function remove(instance) {
    stack = stack.filter(entry => entry !== instance);
    onChange();
  }

  function peek() {
    return stack[stack.length - 1];
  }

  function push(entry) {
    stack.push(entry);
    onChange();
  }

  function onChange() {
    changes.publish(peek());
  }

  return {
    remove,
    peek,
    push,
    subscribe: changes.subscribe
  };
}

var DEFAULT_STACK = createStack();

class Stacked extends _react.default.Component {
  constructor(props) {
    super();

    _defineProperty(this, "state", {
      top: null
    });

    this._unsubscribe = props.stack.subscribe(top => {
      this.setState(() => ({
        top
      }));
    });
  }

  UNSAFE_componentWillMount() {
    var stack = this.props.stack;
    stack.push(this);
  }

  componentWillUnmount() {
    var stack = this.props.stack;

    this._unsubscribe();

    stack.remove(this);
  }

  render() {
    return this.props.children(this.state.top === this);
  }

}

exports.default = Stacked;

_defineProperty(Stacked, "propTypes", {
  children: _propTypes.default.func.isRequired,
  stack: _propTypes.default.shape({
    remove: _propTypes.default.func,
    peek: _propTypes.default.func,
    push: _propTypes.default.func,
    subscribe: _propTypes.default.func
  })
});

_defineProperty(Stacked, "defaultProps", {
  stack: DEFAULT_STACK
});
