// @sanity/components/lib/lists/stories/sortable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SortableStory = SortableStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _react = _interopRequireDefault(require("react"));

var _sortable = require("part:@sanity/components/lists/sortable");

var _reactSortableHoc = require("react-sortable-hoc");

var _chance = _interopRequireDefault(require("chance"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var chance = new _chance.default();
var containerStyle = {
  width: '90%',
  height: '90%',
  boxShadow: '0 0 10px #999',
  overflow: 'hidden',
  position: 'fixed',
  top: '50%',
  left: '50%',
  transform: 'translateX(-50%) translateY(-50%)'
};
var defaultItems = (0, _range2.default)(100).map((item, i) => {
  return {
    key: "".concat(i),
    title: chance.name()
  };
});

class SortableTester extends _react.default.PureComponent {
  constructor(props, args) {
    super(props, args);

    _defineProperty(this, "handleOnSort", event => {
      var items = this.state.items;
      var oldIndex = event.oldIndex,
          newIndex = event.newIndex;
      this.setState({
        items: (0, _reactSortableHoc.arrayMove)(items, oldIndex, newIndex)
      });
      this.props.onSort(event);
    });

    this.state = {
      items: this.props.items.slice()
    };
  }

  render() {
    var items = this.state.items;
    return /*#__PURE__*/_react.default.createElement(_sortable.List, _extends({}, this.props, {
      onSort: this.handleOnSort
    }), items.map((item, index) => {
      return /*#__PURE__*/_react.default.createElement(_sortable.Item, {
        index: index,
        key: String(index)
      }, /*#__PURE__*/_react.default.createElement(_sortable.DragHandle, null), item.title);
    }));
  }

}

function SortableStory() {
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/lists/sortable",
    propTables: [_sortable.List]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: containerStyle
  }, /*#__PURE__*/_react.default.createElement(SortableTester, {
    items: defaultItems
  })));
}
