// @sanity/components/lib/validation/stories/listItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItemStory = ListItemStory;

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

var _ValidationListItem = _interopRequireDefault(require("../ValidationListItem"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ListItemStory() {
  var marker = {
    type: 'validation',
    level: 'error',
    path: ['test'],
    item: {
      paths: [],
      path: ['title'],
      message: (0, _knobs.text)('Message', 'The password must be at least 20 characters, have 5 special characters, minimum 5 lowercase and 2+3 uppercase letters, at least 5 numbers (sequences must not be a part if PI), 2 greek symbols and 5 emojis. Also not use your pet name or birthday of anyone in your family or city.', 'props'),
      name: 'ValidationError'
    }
  };
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: '#fff'
    }
  }, /*#__PURE__*/_react.default.createElement(_ValidationListItem.default, {
    marker: marker,
    path: (0, _knobs.text)('Path', undefined, 'props'),
    truncate: (0, _knobs.boolean)('truncate', false, 'props')
  })));
}
