// @sanity/components/lib/validation/stories/list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListStory = ListStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _status = _interopRequireDefault(require("part:@sanity/components/validation/status"));

var _list = _interopRequireDefault(require("part:@sanity/components/validation/list"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

var _list2 = _interopRequireDefault(require("./list.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

var mockMarkers = length => {
  return (0, _range2.default)(length || 5).map((marker, i) => {
    return {
      type: 'validation',
      level: 'error',
      path: ["test_".concat(i)],
      item: {
        paths: [],
        path: ['title'],
        message: chance.sentence(),
        name: 'ValidationError'
      }
    };
  });
};

function ListStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
    background: false
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/validation/status",
    propTables: [_status.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _list2.default.root
  }, /*#__PURE__*/_react.default.createElement(_list.default, {
    markers: mockMarkers(),
    truncate: (0, _knobs.boolean)('truncate', false, 'props')
  }))));
}
