// @sanity/components/lib/validation/stories/status.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatusStory = StatusStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _status = _interopRequireDefault(require("part:@sanity/components/validation/status"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();
var centerStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100%',
  width: '100%',
  position: 'absolute',
  boxSizing: 'border-box',
  padding: '2rem',
  top: 0,
  left: 0,
  paddingBottom: '10em'
};

var mockMarkers = length => {
  return (0, _range2.default)(length || 5).map((marker, i) => {
    return {
      type: 'validation',
      level: 'error',
      path: ["test_".concat(i)],
      item: {
        paths: [],
        path: ['title'],
        message: chance.sentence(),
        name: 'ValidationError'
      }
    };
  });
};

function StatusStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/validation/status",
    propTables: [_status.default]
  }, /*#__PURE__*/_react.default.createElement(_status.default, {
    markers: mockMarkers(),
    onClick: (0, _actions.action)('onClick')
  })));
}
