// @sanity/form-builder/lib/inputs/PortableText/Objects/renderers/FullscreenObjectEditing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FullscreenObjectEditing = void 0;

var _react = _interopRequireDefault(require("react"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _stacked = _interopRequireDefault(require("part:@sanity/components/utilities/stacked"));

var _presence = require("@sanity/base/presence");

var _FormBuilderInput = require("../../../../FormBuilderInput");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/prop-types */
var FullscreenObjectEditing = (_ref) => {
  var focusPath = _ref.focusPath,
      markers = _ref.markers,
      object = _ref.object,
      onBlur = _ref.onBlur,
      onChange = _ref.onChange,
      onClose = _ref.onClose,
      onFocus = _ref.onFocus,
      path = _ref.path,
      presence = _ref.presence,
      readOnly = _ref.readOnly,
      type = _ref.type;

  var handleChange = patchEvent => onChange(patchEvent, path);

  return /*#__PURE__*/_react.default.createElement(_stacked.default, null, isActive => /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_fullscreen.default, {
    isOpen: true,
    title: type.title,
    onEscape: event => isActive && onClose(event),
    onClose: onClose
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, null, /*#__PURE__*/_react.default.createElement(_FormBuilderInput.FormBuilderInput, {
    type: type,
    level: 0,
    readOnly: readOnly || type.readOnly,
    value: object,
    presence: presence,
    onChange: handleChange,
    onFocus: onFocus,
    onBlur: onBlur,
    focusPath: focusPath,
    path: path,
    markers: markers
  }))))));
};

exports.FullscreenObjectEditing = FullscreenObjectEditing;
