// @sanity/form-builder/lib/inputs/ArrayInput/ConfirmButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _trashIcon = _interopRequireDefault(require("part:@sanity/base/trash-icon"));

var _popover = _interopRequireDefault(require("part:@sanity/components/dialogs/popover"));

var _ConfirmButton = _interopRequireDefault(require("./styles/ConfirmButton.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ConfirmButton extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_button", void 0);

    _defineProperty(this, "_confirmButton", void 0);

    _defineProperty(this, "state", {
      showConfirmDialog: false
    });

    _defineProperty(this, "handleClick", event => {
      this.open();
    });

    _defineProperty(this, "handleClickOutside", event => {
      this.close();
    });

    _defineProperty(this, "handleConfirmPopoverClose", event => {
      this.close();
    });

    _defineProperty(this, "setButton", button => {
      this._button = button;
    });

    _defineProperty(this, "setConfirmButton", button => {
      this._confirmButton = button;
    });
  }

  close() {
    this.setState({
      showConfirmDialog: false
    });
  }

  open() {
    this.setState({
      showConfirmDialog: true
    });
  }

  componentDidUpdate(prevProps, prevState) {
    var wasOpen = prevState.showConfirmDialog;
    var isOpen = this.state.showConfirmDialog;

    if (!wasOpen && isOpen) {
      this._confirmButton.focus();
    } else if (wasOpen && !isOpen) {
      this._button.focus();
    }
  }

  render() {
    var showConfirmDialog = this.state.showConfirmDialog;

    var _this$props = this.props,
        onConfirm = _this$props.onConfirm,
        rest = _objectWithoutProperties(_this$props, ["onConfirm"]);

    return /*#__PURE__*/_react.default.createElement("div", {
      className: _ConfirmButton.default.root
    }, /*#__PURE__*/_react.default.createElement(_default.default, _extends({}, rest, {
      icon: _trashIcon.default,
      onClick: this.handleClick,
      padding: "small",
      ref: this.setButton
    })), showConfirmDialog && /*#__PURE__*/_react.default.createElement(_popover.default, {
      color: "danger",
      useOverlay: false,
      onEscape: this.handleConfirmPopoverClose,
      onClickOutside: this.handleClickOutside,
      padding: "none"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _ConfirmButton.default.wrapper
    }, /*#__PURE__*/_react.default.createElement("div", {
      tabIndex: 0,
      onFocus: this.handleConfirmPopoverClose
    }), /*#__PURE__*/_react.default.createElement(_default.default, {
      color: "danger",
      onClick: onConfirm,
      icon: _trashIcon.default,
      ref: this.setConfirmButton
    }, "Remove?"), /*#__PURE__*/_react.default.createElement("div", {
      tabIndex: 0,
      onFocus: this.handleConfirmPopoverClose
    }))));
  }

}

exports.default = ConfirmButton;
