// @sanity/components/lib/tabs/TabPanel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TabPanel;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _TabPanel = _interopRequireDefault(require("./TabPanel.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function TabPanel(props) {
  var className = [_TabPanel.default.root, props.className].filter(Boolean).join(' ');
  return /*#__PURE__*/_react.default.createElement("div", {
    "aria-labelledby": props['aria-labelledby'],
    className: className,
    id: props.id,
    role: "tabpanel",
    tabIndex: props.tabIndex === undefined ? 0 : props.tabIndex
  }, props.children);
}

TabPanel.propTypes = {
  'aria-labelledby': _propTypes.default.string.isRequired,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  id: _propTypes.default.string.isRequired,
  tabIndex: _propTypes.default.number
};
TabPanel.defaultProps = {
  children: null,
  className: undefined,
  tabIndex: undefined
};
