// @sanity/desk-tool/lib/panes/documentPane/DocumentStatusBar/DocumentStatusBarBadges.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentStatusBarBadges = DocumentStatusBarBadges;

var _reactHooks = require("@sanity/react-hooks");

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/badges/default"));

var _DocumentStatusBarBadges = _interopRequireDefault(require("./DocumentStatusBarBadges.css"));

var _resolver = _interopRequireDefault(require("part:@sanity/base/document-badges/resolver"));

var _utils = require("part:@sanity/base/actions/utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/no-multi-comp */
function DocumentStatusBarBadgesInner(props) {
  if (props.states.length === 0) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: _DocumentStatusBarBadges.default.statusBadges
  }, props.states.map((badge, badgeIndex) => /*#__PURE__*/_react.default.createElement("div", {
    key: String(badgeIndex)
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    color: badge.color,
    title: badge.title
  }, badge.label))));
}

function DocumentStatusBarBadges(props) {
  var editState = (0, _reactHooks.useEditState)(props.id, props.type);
  var badges = editState ? (0, _resolver.default)(editState) : null;
  return badges ? /*#__PURE__*/_react.default.createElement(_utils.RenderBadgeCollectionState, {
    component: DocumentStatusBarBadgesInner,
    badges: badges,
    badgeProps: editState
  }) : null;
}
