// @sanity/desk-tool/lib/panes/documentPane/DocumentStatusBar/HistoryStatusBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HistoryStatusBar = HistoryStatusBar;

var _react = _interopRequireDefault(require("react"));

var _DocumentStatusBar = _interopRequireDefault(require("./DocumentStatusBar.css"));

var _DocumentStatusBarActions = require("./DocumentStatusBarActions");

var _TimeAgo = _interopRequireDefault(require("../../../components/TimeAgo"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function HistoryStatusBar(props) {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _DocumentStatusBar.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _DocumentStatusBar.default.status
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _DocumentStatusBar.default.statusDetails
  }, "Changed ", /*#__PURE__*/_react.default.createElement(_TimeAgo.default, {
    time: props.selectedEvent.endTime
  }), props.isLatestEvent && " (latest)")), /*#__PURE__*/_react.default.createElement("div", {
    className: _DocumentStatusBar.default.actions
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _DocumentStatusBar.default.actionsWrapper
  }, /*#__PURE__*/_react.default.createElement(_DocumentStatusBarActions.HistoryStatusBarActions, {
    id: props.id,
    type: props.type,
    revision: props.selectedEvent.rev
  }))));
}
