// @sanity/components/lib/dialogs/DialogContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _contentStyle = _interopRequireDefault(require("part:@sanity/components/dialogs/content-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DialogContent extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        size = _this$props.size,
        children = _this$props.children,
        padding = _this$props.padding;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)(_contentStyle.default[size], _contentStyle.default["padding_".concat(padding)])
    }, children);
  }

}

exports.default = DialogContent;

_defineProperty(DialogContent, "propTypes", {
  size: _propTypes.default.oneOf(['small', 'medium', 'large', 'auto']),
  padding: _propTypes.default.oneOf(['none', 'small', 'medium', 'large']),
  children: _propTypes.default.node.isRequired
});

_defineProperty(DialogContent, "defaultProps", {
  size: 'auto',
  padding: 'medium'
});
