// @sanity/components/lib/dialogs/PanePopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _circleCheckIcon = _interopRequireDefault(require("part:@sanity/base/circle-check-icon"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _infoIcon = _interopRequireDefault(require("part:@sanity/base/info-icon"));

var _classnames = _interopRequireDefault(require("classnames"));

var _PanePopover = _interopRequireDefault(require("./styles/PanePopover.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class PanePopover extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "DEFAULT_ICONS", {
      info: /*#__PURE__*/_react.default.createElement(_infoIcon.default, null),
      success: /*#__PURE__*/_react.default.createElement(_circleCheckIcon.default, null),
      warning: /*#__PURE__*/_react.default.createElement(_warningIcon.default, null),
      error: /*#__PURE__*/_react.default.createElement(_warningIcon.default, null)
    });

    _defineProperty(this, "iconKind", () => {
      var _this$props = this.props,
          icon = _this$props.icon,
          kind = _this$props.kind;
      if (typeof icon === 'boolean' && icon) return this.DEFAULT_ICONS[kind];
      if (typeof icon === 'object') return icon;
      return undefined;
    });
  }

  render() {
    var _this$props2 = this.props,
        children = _this$props2.children,
        icon = _this$props2.icon,
        id = _this$props2.id,
        kind = _this$props2.kind,
        title = _this$props2.title,
        subtitle = _this$props2.subtitle;
    var Icon = this.iconKind();
    return /*#__PURE__*/_react.default.createElement("div", {
      "aria-label": kind,
      "aria-describedby": "popoverTitle-".concat(kind, "-").concat(id),
      className: (0, _classnames.default)([_PanePopover.default.root, _PanePopover.default.dialog]),
      "data-kind": kind
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.inner
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.content
    }, /*#__PURE__*/_react.default.createElement("div", {
      id: "popoverTitle-".concat(kind, "-").concat(id),
      className: _PanePopover.default.title
    }, icon && /*#__PURE__*/_react.default.createElement("div", {
      role: "img",
      "aria-hidden": true,
      className: _PanePopover.default.icon
    }, Icon), title), subtitle && /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.subtitle
    }, subtitle), children && /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.children
    }, children))));
  }

}

exports.default = PanePopover;

_defineProperty(PanePopover, "propTypes", {
  children: _propTypes.default.node,
  icon: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.bool]),
  kind: _propTypes.default.oneOf(['info', 'warning', 'error', 'success', 'neutral']),
  title: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]).isRequired,
  subtitle: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node]),
  id: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]).isRequired
});

_defineProperty(PanePopover, "defaultProps", {
  children: null,
  icon: null,
  kind: 'info'
});
