// @sanity/desk-tool/lib/panes/documentPane/InspectHistory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _InspectView = _interopRequireDefault(require("./InspectView"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class InspectHistory extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        onClose = _this$props.onClose,
        document = _this$props.document;
    var isLoading = document.isLoading,
        snapshot = document.snapshot;
    return isLoading ? /*#__PURE__*/_react.default.createElement(_spinner.default, null) : /*#__PURE__*/_react.default.createElement(_InspectView.default, {
      value: snapshot,
      onClose: onClose
    });
  }

}

exports.default = InspectHistory;

_defineProperty(InspectHistory, "propTypes", {
  document: _propTypes.default.shape({
    isLoading: _propTypes.default.bool.isRequired,
    snapshot: _propTypes.default.shape({
      _type: _propTypes.default.string
    })
  }).isRequired,
  onClose: _propTypes.default.func.isRequired
});
