// @sanity/desk-tool/lib/panes/documentPane/InspectView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _isObject2 = _interopRequireDefault(require("lodash/isObject"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _reactJsonInspector = _interopRequireDefault(require("react-json-inspector"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _tabList = _interopRequireDefault(require("part:@sanity/components/tabs/tab-list"));

var _tab = _interopRequireDefault(require("part:@sanity/components/tabs/tab"));

var _tabPanel = _interopRequireDefault(require("part:@sanity/components/tabs/tab-panel"));

var _hashlru = _interopRequireDefault(require("hashlru"));

var _reactPropsStream = require("react-props-stream");

var _settings = _interopRequireDefault(require("../../settings"));

var _DocTitle = _interopRequireDefault(require("../../components/DocTitle"));

var _InspectView = _interopRequireDefault(require("./InspectView.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var lru = (0, _hashlru.default)(1000);

function isExpanded(keyPath, value) {
  var cached = lru.get(keyPath);

  if (cached === undefined) {
    lru.set(keyPath, Array.isArray(value) || (0, _isObject2.default)(value));
    return isExpanded(keyPath, value);
  }

  return cached;
}

function toggleExpanded(event) {
  var path = event.path;
  var current = lru.get(path);

  if (current === undefined) {
    // something is wrong
    return;
  }

  lru.set(path, !current);
}

function selectElement(element) {
  var sel = window.getSelection();
  sel.removeAllRanges();
  var range = document.createRange();
  range.selectNodeContents(element);
  sel.addRange(range);
}

function select(event) {
  selectElement(event.currentTarget);
}

function maybeSelectAll(event) {
  var selectAll = event.keyCode === 65 && (event.metaKey || event.ctrlKey);

  if (!selectAll) {
    return;
  }

  event.preventDefault();
  selectElement(event.currentTarget);
}

var VIEW_MODE_PARSED = {
  id: 'parsed',
  title: 'Parsed'
};
var VIEW_MODE_RAW = {
  id: 'raw',
  title: 'JSON'
};
var VIEW_MODES = [VIEW_MODE_PARSED, VIEW_MODE_RAW];

var viewModeSettings = _settings.default.forKey('inspect-view-preferred-view-mode');

function mapReceivedPropsToChildProps(props$) {
  var onViewModeChange = nextViewMode => viewModeSettings.set(nextViewMode.id);

  var viewModeSetting$ = viewModeSettings.listen('parsed').pipe((0, _operators.map)(id => VIEW_MODES.find(mode => mode.id === id)));
  return (0, _rxjs.combineLatest)(props$, viewModeSetting$).pipe((0, _operators.map)((_ref) => {
    var _ref2 = _slicedToArray(_ref, 2),
        props = _ref2[0],
        viewMode = _ref2[1];

    return _objectSpread(_objectSpread({}, props), {}, {
      viewMode,
      onViewModeChange
    });
  }));
}

function InspectView(props) {
  var idPrefix = props.idPrefix,
      onClose = props.onClose,
      onViewModeChange = props.onViewModeChange,
      value = props.value,
      viewMode = props.viewMode; // @todo: prefix with pane id

  var tabIdPrefix = "".concat(idPrefix, "_inspect_");
  return /*#__PURE__*/_react.default.createElement(_fullscreen.default, {
    showHeader: true,
    title: /*#__PURE__*/_react.default.createElement("span", null, "Inspecting", ' ', /*#__PURE__*/_react.default.createElement("em", null, /*#__PURE__*/_react.default.createElement(_DocTitle.default, {
      document: value
    }))),
    onClose: onClose
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
    className: _InspectView.default.toolbar
  }, /*#__PURE__*/_react.default.createElement(_tabList.default, null, /*#__PURE__*/_react.default.createElement(_tab.default, {
    "aria-controls": "".concat(tabIdPrefix, "tabpanel"),
    id: "".concat(tabIdPrefix, "tab-").concat(VIEW_MODE_PARSED.id),
    isActive: viewMode === VIEW_MODE_PARSED,
    label: "Parsed" // eslint-disable-next-line react/jsx-no-bind
    ,
    onClick: () => onViewModeChange(VIEW_MODE_PARSED)
  }), /*#__PURE__*/_react.default.createElement(_tab.default, {
    "aria-controls": "".concat(tabIdPrefix, "tabpanel"),
    id: "".concat(tabIdPrefix, "tab-").concat(VIEW_MODE_PARSED.id),
    isActive: viewMode === VIEW_MODE_RAW,
    label: "Raw JSON" // eslint-disable-next-line react/jsx-no-bind
    ,
    onClick: () => onViewModeChange(VIEW_MODE_RAW)
  }))), /*#__PURE__*/_react.default.createElement(_tabPanel.default, {
    "aria-labelledby": "".concat(tabIdPrefix, "tab-").concat(viewMode.id),
    className: _InspectView.default.content,
    id: "".concat(tabIdPrefix, "tabpanel"),
    role: "tabpanel"
  }, viewMode === VIEW_MODE_PARSED && /*#__PURE__*/_react.default.createElement("div", {
    className: _InspectView.default.jsonInspectorContainer
  }, /*#__PURE__*/_react.default.createElement(_reactJsonInspector.default, {
    isExpanded: isExpanded,
    onClick: toggleExpanded,
    data: value
  })), viewMode === VIEW_MODE_RAW && /*#__PURE__*/_react.default.createElement("pre", {
    className: _InspectView.default.raw,
    tabIndex: 0,
    onKeyDown: maybeSelectAll,
    onDoubleClick: select,
    onFocus: select
  }, JSON.stringify(value, null, 2)))));
}

InspectView.propTypes = {
  idPrefix: _propTypes.default.string.isRequired,
  onClose: _propTypes.default.func,
  onViewModeChange: _propTypes.default.func.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  value: _propTypes.default.object,
  viewMode: _propTypes.default.shape({
    id: _propTypes.default.string.isRequired,
    title: _propTypes.default.string.isRequired
  }).isRequired
};
InspectView.defaultProps = {
  onClose: undefined,
  value: undefined
};

var _default = (0, _reactPropsStream.withPropsStream)(mapReceivedPropsToChildProps, InspectView);

exports.default = _default;
